### PR TITLE
feat: New message composer

### LINKED ISSUE #5301: User suggestions not available when using a slash command
### Description:

When composing a message, no usernames will be suggested when starting the message with a slash command.

### Environment Information:

- Rocket.Chat Server Version: 6.4.3
- Rocket.Chat App Version: 4.42.0.50842
- Device Name: ELE-L29 (Huawei P30)
- OS Version: EMUI 12.0.0 (Android 11)

### Steps to reproduce:

1. Start composing a message with a slash command, like `/me`, `/shrug` etc.
2. After a space, type `@` and wait for username suggestions to appear.

### Expected behavior:

Username suggestions are shown.

### Actual behavior:

No usernames are suggested.

https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/404840/f0c46945-54ad-45c6-aa30-dbc9644f26b4

### Additional context:

In the browser version of Rocket.Chat, this works as expected.

### PR DESCRIPTION
**Pull request by diegolmello**
## Issue(s)	

Closes #5301

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/containers/MessageComposer/MessageComposer.tsx [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+					text.substr(0, (cursor > 0 ? cursor : text.length) - charsToRemove) + text.substr(cursor > 0 ? cursor : text.length);
+				newCursor = cursor - charsToRemove;
+				composerInputComponentRef.current.setInput(newText, { start: newCursor, end: newCursor });
+				break;
+			case EventTypes.EMOJI_PRESSED:
+				// logEvent(events.MB_EMOJI_SELECTED);
+				let emojiText = '';
+				if (typeof emoji === 'string') {
+					const shortname = `:${emoji}:`;
+					emojiText = shortnameToUnicode(shortname);
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/MessageComposer.tsx:158]: Really good change

File: app/containers/MessageComposer/MessageComposer.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+				} else {
+					emojiText = `:${emoji.name}:`;
+				}
+				newText = `${text.substr(0, cursor)}${emojiText}${text.substr(cursor)}`;
+				newCursor = cursor + emojiText.length;
+				composerInputComponentRef.current.setInput(newText, { start: newCursor, end: newCursor });
+				break;
+			case EventTypes.SEARCH_PRESSED:
+				// logEvent(events.MB_EMOJI_SEARCH_PRESSED);
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/MessageComposer.tsx:167]: I have a task to evaluate and add all events

File: app/containers/MessageComposer/components/Autocomplete/Autocomplete.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+	if (items.length === 0 || !type) {
+		return null;
+	}
+
+	if (type !== '/preview') {
+		return (
+			<View
+				style={{
+					maxHeight: 216,
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/components/Autocomplete/Autocomplete.tsx:33]: Changed to make it more clear. Take a look.
Hopefully we fix them soon 👀

File: app/containers/MessageComposer/components/Autocomplete/Autocomplete.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+					left: 8,
+					right: 8,
+					backgroundColor: colors.surfaceNeutral,
+					position: 'absolute',
+					bottom: trackingViewHeight + keyboardHeight + 50,
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/components/Autocomplete/Autocomplete.tsx:38]: Same as above

File: app/containers/MessageComposer/components/Autocomplete/AutocompleteCannedResponse.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import I18n from '../../../../i18n';
+import { CustomIcon } from '../../../CustomIcon';
+import { NO_CANNED_RESPONSES } from '../../constants';
+
+export const AutocompleteCannedResponse = ({ item }: { item: IAutocompleteCannedResponse }) => {
+	const { colors } = useTheme();
+	if (item.id === NO_CANNED_RESPONSES) {
+		return (
+			<>
+				<View style={{ flex: 1, justifyContent: 'center' }}>
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/components/Autocomplete/AutocompleteCannedResponse.tsx:16]: I'm doing all of them rn.

File: app/containers/MessageComposer/context.tsx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+			setMicOrSend,
+			setMarkdownToolbar,
+			setAlsoSendThreadToChannel,
+			setRecordingAudio,
+			setAutocompleteParams
+		};
+	}, []);
+
+	return (
+		<MessageComposerContextApi.Provider value={api}>
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/MessageComposer/context.tsx:169]: Got it from this article https://www.developerway.com/posts/how-to-write-performant-react-apps-with-context
It's strange at first, but it's way more performant and straight forward to maintain.

File: jest.setup.js [hunk 7]
```diff
@@ -1,5 +1,5 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply jest.setup.js:1]: Temporary. It was creating ts issues for no clear reason and I'm going to write tests only now.

File: app/i18n/locales/zh-TW.json [hunk 8]
```diff
@@ -489,5 +489,27 @@
...
+  "Invite_user_to_join_channel": "邀請使用者加入此頻道",
+  "Invite_user_to_join_channel_all_from": "邀請[#channel]中的所有使用者加入此頻道",
+  "Invite_user_to_join_channel_all_to": "邀請此頻道的所有使用者加入[＃頻道]",
+  "Join_the_given_channel": "加入現有頻道",
+  "Remove_someone_from_room": "從房間中刪除某人",
+  "Leave_the_current_channel": "離開目前頻道",
+  "Displays_action_text": "顯示文字行動",
+  "Direct_message_someone": "私訊其他人",
+  "Mute_someone_in_room": "在房間中有人靜音",
+  "Unmute_someone_in_room": "有人在房間取消靜音",
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/i18n/locales/zh-TW.json:513]: They are needed for some slash commands to work. Got them from frontend repo.

File: app/containers/EmojiPicker/Footer.tsx [hunk 9]
```diff
@@ -6,15 +6,12 @@ import { CustomIcon } from '../CustomIcon';
...
 
-const BUTTON_HIT_SLOP = { top: 15, right: 15, bottom: 15, left: 15 };
-
 const Footer = ({ onSearchPressed, onBackspacePressed }: IFooterProps): React.ReactElement => {
 	const { colors } = useTheme();
 	return (
 		<View style={[styles.footerContainer, { borderTopColor: colors.borderColor }]}>
 			<Pressable
 				onPress={onSearchPressed}
-				hitSlop={BUTTON_HIT_SLOP}
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/containers/EmojiPicker/Footer.tsx:17]: 44x44 is big enough, so it doesn't need a hit slop imo

**Review by GleidsonDaniel — APPROVED** [review]

**Comment by GleidsonDaniel** [comment]
bug when inserting emojis within multiple characters:
https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/18687097/eee8954c-1d12-471d-a091-6a1f9ea71c03

**Comment by GleidsonDaniel** [comment]
This is not a bug, however this "jump" that occurs when clicking on the input could be avoided if you disabled the input and only focused when clicking on it.
Like, before it goes up it is disabled and then when you click on the input it goes up and focuses on it.
https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/18687097/21c84cd1-2d9b-4d0a-a068-2f6551ebf9d2

**Pull request merged**